src/app/address/controllers/address/nonUKAddress.js

```diff
@@ -0,0 +1,118 @@
+const BaseController = require("hmpo-form-wizard").Controller;
+const {
+  validateBuildingAddress,
+} = require("../../validators/nonUKAddressValidator");
+const { yearFrom, getCountry } = require("../../../../lib/helpers");
+
+class NonUKAddressController extends BaseController {
+  getValues(req, res, callback) {
+    super.getValues(req, res, (err, values) => {
+      values.addressCountryName = this.mapToName(
+        req.translate,
+        req.sessionModel.get("country")
+      );
+
+      Object.entries(req?.form?.errors || {})?.map(
+        ([fieldName, validationAttribute]) =>
+          (values[`${fieldName}InValid`] = this.getErrorMessage(
+            req.translate,
+            validationAttribute.key,
+            validationAttribute.type
+          ))
+      );
+      // Display the buildAddress error message only when all input fields are empty
+      values.buildAddressErrorMessage = this.hasBuildingEmptyErrors(req) && {
+        text: req.translate("validation.apartmentOrBuildingNumberOrName"),
+        visuallyHiddenText: "error",
+      };
+
+      callback(err, values);
+    });
+  }
+
+  validateFields(req, res, callback) {
+    const formFields = req.form.options.fields;
+    const apartment = "nonUKAddressApartmentNumber";
+    const number = "nonUKAddressBuildingNumber";
+    const name = "nonUKAddressBuildingName";
+
+    const buildingAddress = [
+      [apartment, req.body[apartment]],
+      [number, req.body[number]],
+      [name, req.body[name]],
+    ];
+
+    buildingAddress.every(fieldIsEmpty) &&
+      this.defaultToFirstField(formFields, apartment, req);
+
+    super.validateFields(req, res, callback);
+  }
+
+  async saveValues(req, res, callback) {
+    super.saveValues(req, res, () => {
+      const addressCountry = req.sessionModel.get("country");
+      const address = this.buildAddress(req.body, addressCountry);
+      req.sessionModel.set("address", address);
+
+      callback();
+    });
+  }
+
+  buildAddress(
+    {
+      nonUKAddressApartmentNumber,
+      nonUKAddressBuildingNumber,
+      nonUKAddressBuildingName,
+      nonUKAddressStreetName,
+      nonUKAddressLocality,
+      nonUKAddressPostalCode,
+      nonUKAddressRegion,
+      nonUKAddressYearFrom,
+    },
+    addressCountry
+  ) {
+    return {
+      addressRegion: nonUKAddressRegion,
+      addressLocality: nonUKAddressLocality,
+      streetName: nonUKAddressStreetName,
+      postalCode: nonUKAddressPostalCode,
+      buildingNumber: nonUKAddressBuildingNumber,
+      buildingName: nonUKAddressBuildingName,
+      subBuildingName: nonUKAddressApartmentNumber,
+      validFrom: yearFrom(nonUKAddressYearFrom),
+      addressCountry,
+    };
+  }
+
+  mapToName(translate, countryCode) {
```

Comment:
Did we decide that this isn't going to be needed on the confirm page?

Reply:
Yeah, but I haven't seen this pattern used anywhere in this repo, the templates have been made as dumb as possible so I think we should stick to having the logic in the backend.

Reply:
Hmm keen to hear more about why you feel like this is logic?

From my perspective I think we should be saving the translation key to the value (`values.addressCountryName = getCountry(countryCode).key`) and then doing the translation in the template like we do for all other strings - eg in the [njk file you've added](https://github.com/govuk-one-login/ipv-cri-address-front/pull/1008/files#diff-0e0d6974cf2013c819c61f5faff714aebe704ea8fff74faf0fc65e6cd2bf7526R10)

Reply:
Hmm, I would have to re-work `changeCountry`

```
links:
  changeAddress: <a href="/address/edit?edit=true" id="change-address" data-id="changeAddress" class="govuk-link">Change<span class="govuk-visually-hidden"> your current address</span></a>
  changeYearFrom: <a href="/address/edit?edit=true" id="change-address" data-id="changeAddress" class="govuk-link">Change<span class="govuk-visually-hidden"> the year you started living at your current address</span></a>
  changePostcode: <p>Postcode <br> <span data-id="changePostcodeValue"><b>{{values.addressPostcode}}</b></span> &nbsp;&nbsp;&nbsp;&nbsp; <a href="/search" data-id="changePostcode" class="govuk-link">Change<span class="govuk-visually-hidden"> your current postcode</span></a></p>
  changeCountry: <p>Country <br> <span data-id="changeCountryValue"><b>{{values.addressCountryName}}</b></span> &nbsp;&nbsp;&nbsp;&nbsp; <a href="/what-country" data-id="changeCountry" class="govuk-link">Change<span class="govuk-visually-hidden"> your current country</span></a></p>
```

Reply:
What about this would have to change?